QueryCiv3/Bic.cs

```diff
@@ -9,9 +10,253 @@ public class BicData
+                                        TerrGood[i,j] = Util.GetFlag(*(bytePtr + offset + i * terrLen + 8 + (j / 8)), j % 8);
+                                    }
+                                    // Copy remaining bytes into TERR struct:
+                                    Buffer.MemoryCopy(bytePtr + offset + (i + 1) * terrLen - sizeof(TERR) + 8, terrBytePtr + 8, sizeof(TERR) - 8, sizeof(TERR) - 8);
+                                    terrBytePtr += sizeof(TERR);
+                                }
+                            }
+                            break;
+                        case "TFRM":
+                            dataLength = count * sizeof(TFRM);
+                            Tfrm = new TFRM[count];
+                            fixed (void* ptr = Tfrm) {
+                                Buffer.MemoryCopy(bytePtr + offset, ptr, dataLength, dataLength);
+                            }
+                            break;
+                        case "TILE":
+                            dataLength = count * sizeof(TILE);
+                            Tile = new TILE[count];
+                            fixed (void* ptr = Tile) {
+                                Buffer.MemoryCopy(bytePtr + offset, ptr, dataLength, dataLength);
+                            }
+                            break;
+                        case "UNIT":
+                            dataLength = count * sizeof(UNIT);
+                            Unit = new UNIT[count];
+                            fixed (void* ptr = Unit) {
+                                Buffer.MemoryCopy(bytePtr + offset, ptr, dataLength, dataLength);
+                            }
+                            break;
+                        case "WCHR":
+                            dataLength = count * sizeof(WCHR);
+                            Wchr = new WCHR[count];
+                            fixed (void* ptr = Wchr) {
+                                Buffer.MemoryCopy(bytePtr + offset, ptr, dataLength, dataLength);
+                            }
+                            break;
+                        case "WMAP":
+                            dataLength = -7;
+                            break;
+                        case "WSIZ":
+                            dataLength = count * sizeof(WSIZ);
+                            Wsiz = new WSIZ[count];
+                            fixed (void* ptr = Wsiz) {
+                                Buffer.MemoryCopy(bytePtr + offset, ptr, dataLength, dataLength);
+                            }
+                            break;
+                        default:
+                            // Once every BIQ section is set up with the correct associate struct(s), the default case where a header isn't found
+                            // should never occur. However, for now, if a header isn't found, we'll just increment by 1 to keep searching
+                            dataLength = -7; // We added 8 earlier, so subtract 7
```

Comment:
Okay, I see the 7 here with a comment.  But after reading the comment (and through the offset += dataLength, I am fairly sure that this is not the right way to skip ahead.

We do indeed want to keep searching here.  Line 257 adds dataLength to offset, to continue on to the next section.  For most sections (let's choose World Size as an example), that will change it by the size of the section, times the number of instances of the section, and as you note we've already added 8 for the section header/# of instances.

But let's say we don't know about WSIZ yet, and there are 5 world sizes.  We've added 8 for WSIZ and the int 5.  We subtract 7 here.  So the next time, it's going to see "SIZ\0 ", i.e. SIZ plus a null byte from the first byte of the integer.

We don't actually want it to jump ahead by one byte though.  We should jump ahead of the entire WSIZ section.

To do that, we can leave the 8, but we also need to, for each instance of WSIZ, read the "length of section" integer (which could vary for each instance), and add that many bytes to the offset.

It's a little more complex, but should be considerably more efficient than checking each byte.  It also avoids potential errors if one of our expected headers shows up in the data of another section.  Which isn't likely, but the advantage of this header-iteration approach over just searching as we did before is that it avoids any possibility of failure for that reason.

Reply:
Like a mentioned in one of the above comments, the -7 thing is only temporary and won't be an issue once all sections have their data population. For your WSIZ example, the 7 actually isn't involved at all. When the first WSIZ header is encountered, the offset adds 8 like you say to skip past the header and length. But then the next offset change is for the entire WSIZ section and all its members. So there isn't any byte searching at all: the offset jumps straight to the location of whatever the next header is. Linear searching (where the -7 + 8 = 1 is involved) only begins to happen currently when a header is encountered that we don't yet have the struct(s) set up for, so we don't know how far to jump ahead for the next section. But once they're all set up (hopefully I can get them all in tomorrow), linear byte searching can be removed entirely.

Reply:
Right, it's the linear searching where this would occur, and my thoughts were related to the case where we don't know a header, and thus would (currently) linearly search.  My thinking was that even if we don't know the contents of a section, we know enough to be able to jump ahead in increments of _dataLength_ bytes (as _dataLength_ is always the first 4 bytes of any section, which would be much more efficient, and safer.

But if we're going to have all the rest of the sections set up this week, it's not worth optimizing that in the meantime.  That's probably the best part of this technique, it lets the basic structure get set up quite quickly.

Marking as resolved since this will be obsolete soon.